### PR TITLE
feat: add health probes to node & go service

### PR DESCRIPTION
**Pull request by lance**
Until the next release of Quarkus in November, we'll need to have some special
checks so that we don't add these probes to Quarkus based functions. In the
meantime, we can at least set the liveness and readiness endpoints on the
Node.js/Go functions to /health/liveness and /health/readiness respectively.

Fixes: https://github.com/boson-project/faas/issues/169

Signed-off-by: Lance Ball <lball@redhat.com>

### CODE REVIEW

**Comment by matejvasek** [comment]
@lance I am getting port 0 instead 8080 in the resulting `ksvc`.

**Comment by matejvasek** [comment]
```yaml
         livenessProbe:
           httpGet:
             path: /health/liveness
             port: 0
         name: user-container
         readinessProbe:
           httpGet:
             path: /health/readiness
             port: 0
           successThreshold: 1
```

**Comment by matejvasek** [comment]
I tired fix that and got:
```
Error: knative deployer failed to deploy the service: admission webhook "validation.webhook.serving.knative.dev" denied the request: validation failed: must not set the field(s): spec.template.spec.containers[0].livenessProbe.httpGet.port, spec.template.spec.containers[0].readinessProbe.httpGet.port
```

File: knative/deployer.go [hunk 1]
```diff
@@ -90,7 +91,20 @@ func (d *Deployer) Deploy(f faas.Function) (err error) {
 	return nil
 }
 
-func generateNewService(name, image string) *servingv1.Service {
+func probeFor(port, url string) *corev1.Probe {
+	return &corev1.Probe{
+		Handler: corev1.Handler{
+			HTTPGet: &corev1.HTTPGetAction{
+				Path: url,
+				Port: intstr.IntOrString{
```
Comment (matejvasek) [inline knative/deployer.go:99]: `intstr.FromString(port)` might be better.
But leads to validation error mentioned before.

**Comment by matejvasek** [comment]
Maybe the port shouldn't be set for `ksvc`. Maybe the knative operator sets it for `pods` automatically.
```
 420   │     livenessProbe:
 421   │       failureThreshold: 3
 422   │       httpGet:
 423   │         httpHeaders:
 424   │         - name: K-Kubelet-Probe
 425   │           value: queue
 426   │         path: /health/liveness
 427   │         port: 8012
 428   │         scheme: HTTP

 536   │     readinessProbe:
 537   │       exec:
 538   │         command:
 539   │         - /ko-app/queue
 540   │         - -probe-period
 541   │         - "0"
 542   │       failureThreshold: 3
 543   │       periodSeconds: 1
 544   │       successThreshold: 1
 545   │       timeoutSeconds: 10



```

**Comment by matejvasek** [comment]
@rhuss  could you pleas help us here, how are readiness/liveness probes supposed to work with knative service?

**Comment by lance** [comment]
@matejvasek your function was deployed when you first tried with the PR as written, and it came up? If so, then they work as expected, right? I tested on a local cluster and it was successful for me. When testing with `Port` removed, the service failed to come up.

**Comment by matejvasek** [comment]
@lance 
I think it should be:
```diff
diff --git a/knative/deployer.go b/knative/deployer.go
index 5745ed0..760e51e 100644
--- a/knative/deployer.go
+++ b/knative/deployer.go
@@ -8,7 +8,6 @@ import (
        corev1 "k8s.io/api/core/v1"
        "k8s.io/apimachinery/pkg/api/errors"
        metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
-       "k8s.io/apimachinery/pkg/util/intstr"
        servinglib "knative.dev/client/pkg/serving"
        "knative.dev/client/pkg/wait"
        servingv1 "knative.dev/serving/pkg/apis/serving/v1"
@@ -91,14 +90,11 @@ func (d *Deployer) Deploy(f faas.Function) (err error) {
        return nil
 }
 
-func probeFor(port, url string) *corev1.Probe {
+func probeFor(url string) *corev1.Probe {
        return &corev1.Probe{
                Handler: corev1.Handler{
                        HTTPGet: &corev1.HTTPGetAction{
                                Path: url,
-                               Port: intstr.IntOrString{
-                                       StrVal: port,
-                               },
                        },
                },
        }
@@ -115,8 +111,8 @@ func generateNewService(name, image string, healthChecks bool) *servingv1.Servic
        }
 
        if healthChecks {
-               containers[0].LivenessProbe = probeFor("8080", "/health/liveness")
-               containers[0].ReadinessProbe = probeFor("8080", "/health/readiness")
+               containers[0].LivenessProbe = probeFor("/health/liveness")
+               containers[0].ReadinessProbe = probeFor("/health/readiness")
        }
 
        return &v1.Service{

```

With that it works for me (service is up).

This:
```go
Port: intstr.IntOrString{
    StrVal: port,
}
```
doesn't hurt, but it does nothing, because there is `Type` member it the struct and by default it's `Int`, so setting just `StrVal` doesn't have any effect.

**Comment by matejvasek** [comment]
>When testing with Port removed, the service failed to come up.

that's odd, it works for me, I think removing it should have no effect.

File: knative/deployer.go (hunk 1 above)
Comment (markusthoemmes) [reply knative/deployer.go:99]: You should be able to drop this completely. The "correct" port is inferred. The readiness endpoint has to be served on the "default" server though (the one listening on `$PORT`).

**Review by matejvasek — APPROVED** [review]

**Pull request merged**